datastore/fsds.nim

```diff
@@ -217,6 +219,18 @@ method query*(
   iter.next = next
   return success iter
 
+method modifyGet*(
+  self: FSDatastore,
+  key: Key,
+  fn: ModifyGet): Future[?!seq[byte]] =
+  defaultModifyGetImpl(self, self.lock, key, fn)
```

Comment:
Hmm... as far as I can tell, this will lock the entire datastore regardless of which key is being accessed. I've been thinking about locking for the stores for some time and I think the best approach would be to have a lock per key.

One way of doing this would be to have a table with keys and use the `mgetOrPut` method, like so:

```nim
   try:
     let lock = self.locks.mgetOrPut(key, newAsyncLock())
     defaultModifyGetImpl(self, self.lock, key, fn)
   finally:
     if not lock.locked:
        self.locks.del(key)
```

But thinking about this some more, perhaps, we don't want to handle locking inside the `modifyGet` itself and leave it to the store, since it has a lot more context to make the appropriate decisions, some stores don't even require locking, for example sqlite handles locking internally.

Reply:
I fully agree that locking per key is better than locking per entire ds, I think I oversimplified the default implementation. I used your code and it's working well and tests are passing.

>But thinking about this some more, perhaps, we don't want to handle locking inside the `modifyGet` itself and leave it to the store

It is already the case, `modify` and `modifyGet` are just interface methods. The code that is locking is just a `defaultModifyGetImpl` that can be used or not.